---
title: "Introduction of coADAPTr"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction of coADAPTr}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(coADAPTr)
```

# This is an introduction to coADAPTr (Covalent Labeling Automated Data Analysis Platform in R)

**Introduction to coADAPTr**

This document provides an overview of the coADAPTr package, which stands
for Covalent Labeling Automated Data Analysis Platform in R. coADAPTr is
designed to assist in the processing of raw mass spectrometry data that
has been sequence searched.

**Overview**

The package offers a series of functions to streamline the data analysis
workflow, from importing raw data to generating visualizations and
saving results.

The initial input is an excel sheet (.xlsx or .xls) that contains the
minimum columns: - Master Protein Accessions - Modifications
(Modification Type; Modification Position - Sequence or Peptide (If
using Proteome Discoverer 2.5\> be sure to change "Annotated Sequence'
to "Sequence") - Precursor Abundance/Intensity - Spectrum File (Or any
columns that can be used to separate sample types such as Experimental
Condition and Laser vs No Laser).

If using PD 2.0-3.0 please import all data and run the remove_dup()
function to get ride of duplicates created by the multilevel sequence
searching method described in [Rinas et al.
2016](https://link.springer.com/article/10.1007/s00216-016-9369-3 "An efficient quantitation strategy for hydroxyl radical-mediated protein footprinting using Proteome Discoverer").

*Do not use the function remove_dup if your data comes from another
source or the duplicates have already been removed in excel.*

## Analyzing Sequence Searched Data with coADAPTr

*Before diving into the analysis, it's essential to ensure that your
sequence searched data is in an excel.This package only supports excel
inputs not csv.*

## Step-by-Step Guide for LFQ Data Analysis

Please see this video where coADAPTR was used to analyze LFQ data to
compliment this step-by-step guide

### Step 1: Load Necessary Packages and Data

Before Analyzing your data be sure to run this function to install
necessary packages, resolve package conflicts, import required data, and
select result output folders. ***Follow the console prompts and
anticipate File Explorer pop ups.***

**setup()**

[To elaborate, the following functions are wrapped into one sequence
which is called setup().]{.underline}

1.  before_beginning()--Installs required packages and resolved
    conflicts.
2.  raw_data \<\<- import_data()--Allows the user to select (via file
    explorer) and import the excel file containing their data.
3.  file_output \<\<- output_folder()--Allows the user to select (via
    file explorer) the folder they want the data stored in.
4.  FASTA \<\<- FASTA_file()--Allows the user to select (via file
    explorer) and import the FASTA file in .fasta format of which the
    data was searched.

### Step 2: Prepare the LFQ Data for Analysis

Prepare your LFQ data for analysis by selecting the relevant data,
renaming the columns appropriately, and searching the FASTA file against
peptide spectral matches to identify the start and end residues.
***Follow the console prompts and anticipate File Explorer pop ups.***

**LFQ_Prep()**

[To elaborate, the following functions are wrapped into one sequence
which is called LFQ Prep().]{.underline}

1.  selected_data \<\<- column_selectionLFQ(raw_data)--Prompts the user
    to select the columns corresponding to the request.
2.  modified_data \<\<- rename_and_split_spectrum_files(selected_data)
    --Prompts the user to select and rename the column containing
    Spectrum File or a similar identifier in the format of
    **Condition:Sample Type & \#**. (*ex. Drug Treated:NL1, Drug
    Free:L3, or Outer Region:L2*). Then the Spectrum file column is
    split so Condition is in one column leaving sample type information
    in the Spectrum File column.
3.  modified_data \<\<- SampleControl(modified_data)--Detects sample vs
    control files based on the Spectrum File Column.
4.  modified_data_annotated \<\<-
    annotate_features(modified_data)--Identifies the modification
    position on the peptide based off of input in modification column.
5.  FASTA\<\<- parse_fasta(FASTA)--Parses the FASTA file for relevant
    information to map the modification on the protein sequence.
6.  mod_data_fasta_merged\<\<-
    locate_startend_res(modified_data_annotated, FASTA)--Maps the
    modification to the protein sequence after merging the FASTA file
    with the data.

### Step 3: Calculate the Extent of Modification

Calculate the Extent of Modification (EOM) for each peptide and residue
based on the LFQ data.

**EOM_Calculations()**

*The data frame **quant_graph_df** contains all of the data that is
acceptable for graphing.*

*The data frame **Areas_pep** and **Areas_res** contain the data that is
used to calculate the EOM and will contain all data. This includes cases
where the EOM could be negative (high background oxidation) or the SD is
greater than the EOM (data has a high variance-likely due to
experimental conditions)*

[To elaborate, the following functions are wrapped into one sequence
which is called EOM_Calculations().]{.underline}

Areas_pep\<\<- area_calculations_pep(mod_data_fasta_merged)--Calculates
the Extent of Modification (EOM) and SD (Variance) at the peptide level

graphing_df_pep\<\<- merge_metadata_pep(Areas_pep,
mod_data_fasta_merged)-- Merges the metadata with numeric graphing data

quant_graph_df_pep\<\<-
filtered_graphing_df_pep(graphing_df_pep)--Filters peptide level
graphical data to include data that is acceptable

Areas_res\<\<- area_calculations_res(mod_data_fasta_merged)--Calculates
the Extent of Modification (EOM) and SD (Variance) at the residue level

graphing_df_res\<\<- graphing_data_res(Areas_res,
mod_data_fasta_merged)--Subsets sequence metadata like residue
start/stop for residue level data

quant_graph_df_res\<\<-
filtered_graphing_df_res(graphing_df_res)--Filter residue level
graphical data to include data that is acceptable

### Step 4: Save ALL Tables and Graphs 

Save data frames as Excel files and save grouped bar from the plots as
PNG files. ***Follow the console prompts and anticipate File Explorer
pop ups.***

Tables_and_Graphs()

### Step 5: Remove Duplicate Entries

## Step-by-Step Guide for TMT Data Analysis

### Step 1: Read in Required Inputs

### Step 2: Set Output Directory

### Step 3: Read in Data from Proteome Discoverer

### Step 4: Identify Sample and Control Spectrum Files

### Step 5: Remove Duplicate Entries

### Step 6: Clean and Parse Data

### Extra: Create and Save Venn Diagrams

Create and save venn diagrams using this function:

venn_diagram()

The excel workbook containing the information to place in the venn
diagram have a a column list of master protein accessions correlation to
a condition that is indicated in the header.

![](images/venn%20data.PNG){width="462"}

### 

### 

Annotate the features in the data from the PD output file like
extracting the modification position using the annotate_features()
function:

pd_data_annotated \<- annotate_features(pd_data)

Input your raw data containing the SampleControl column into this
function

![](images/annotated.PNG)

### Step 7: Parse the FASTA File

Parse the FASTA file to extract relevant information and rename the
column headers using the parse_fasta() function:

FASTA \<- parse_fasta(FASTA)

![](images/FASTA.PNG)

Input your FASTA file into this function

### Step 8: Locate Residue Numbers

Locate the residue numbers in the FASTA file using the
locate_startend_res() function:

pd_data_fasta_merged \<- locate_startend_res(pd_data_annotated)

![](images/locate.PNG)

Input your annotated raw data into this function.

**Step 9: Calculate Peptide Level Areas and Extent of Modification**

Calculate the peptide level areas and extent of modification using the
area_calculations_pep() function:

Areas_pep \<- area_calculations_pep(pd_data_fasta_merged)

Input your annotated data merged with your FASTA file into this
function.

### Step 10: Grab Sequence Meta Data (SKIP)

YOU DO NOT NEED TO MANUALLY DO THIS STEP AS IT USED IN THE NEXT FUNCTION

### Step 11: Merge Metada with Graphing Data

Merge the metadata (ie. the peptide and modification location) with the
graphing data using the merge_metadata() function:

graphing_df_pep \<- merge_metadata_pep(Areas_pep, pd_data_fasta_merged)

Input your data containing the peptide level EOM and the raw data that
was merged with the FASTA file into this function.

### Step 12: Filter Graphical Data

Filter the graphical data to include only acceptable data using the
filtered_graphing_df_pep() function:

quant_graph_df_pep \<- filtered_graphing_df_pep(graphing_df_pep)

Input your peptide level graphing data into this function.

### Step 13: Calculate Residue Level Areas and Extent of Modification

Calculate the residue level areas and extent of modification using the
area_calculations_res() function:

Areas_res \<- area_calculations_res(pd_data_fasta_merged)

Input your raw data merged with your FASTA file into this function.

### Step 14: Prepare Residue Level Graphing Data

Prepare the residue level graphing data by annotating the peptide and
residue location using the graphing_data_res() function:

graphing_df_res \<- graphing_data_res(Areas_res, pd_data_fasta_merged)

Input your containing the residue EOM and the raw data that was merged
with the FASTA file into this function.

### Step 15: Filter Residue Level Graphing Data

Filter the residue level graphing data for quantifiable modifications
using the filtered_graphing_df_res() function:

quant_graph_df_res \<- filtered_graphing_df_res(graphing_df_res)

Input your graphing residue level data into this function.

### Step 16: Save Tables and Plots

#### 16A: Create and Save the Tables and Plots Generated During the Analysis

TotalsTable\<- create_totals_tablelist (quant_graph_df_pep,
quant_graph_df_res)

![](images/totals%20table.PNG){width="579"}

#### 16B: Save the Data Frames Generated During Analysis

save_data_frames(file_output, TotalsTable = TotalsTable,
quant_graph_df_pep = quant_graph_df_pep, quant_graph_df_res =
quant_graph_df_res, graphing_df_pep = graphing_df_pep, graphing_df_res =
graphing_df_res)

To save a data frame that is not listed simply add (dataframe =
dataframe) to the end of the comma separated list.

![](images/saved%20data%20frames.PNG){width="396"}

#### 16C: Generate Peptide Level Bar Graphs

To generate the peptide level bar graphs use this function:

generate_eom_plot_pep(df_in = quant_graph_df_pep, file_output =
file_output)

![](images/P10809.png){width="475"}

#### 16D: Generate Residue Level Bar Graphs

To generate the residue level bar graphs use generate_eom_plot_res():

generate_eom_plot_res(df_in = quant_graph_df_res, file_output =
file_output)

![](images/P10809-01.png){width="528"}

### Step 18: Creating and Saving Grouped Bar Graphs.

Once you have analyzed data of different conditions, combine the
quantifiable graphing data into one data frame. In the last column
indicate which row corresponds to which condition. Then use this
function to create your grouped bar plot.

#### 18A Generate Peptide Level Grouped Bar Plot

grouped_bar_plot_pep()

The excel workbook containing the extent of modification for varying
conditions should look like this.

Once you analyzed data o difference conditions simply add a column
"Condition" as a column in excel sheet containing quantifiable
modifications. Then you can combine the two data frames together for
graphing.

![](images/grouped%20bar%20graph%20data.PNG){width="572"}

#### 18B Generate Residue Level Grouped Bar Plot

For residue level EOM grouped bar graphs use the following function:

grouped_bar_plot_res()

This vignette includes explanations and documentation for each step in
the data analysis workflow. It should serve as a comprehensive guide for
users of the `coADAPTr` package.
